hphp/runtime/vm/jit/vasm-lower-inl.h

```diff
@@ -38,14 +44,23 @@ void vasm_lower(Vunit& unit, Vlower lower_impl) {
   // iterators.
   auto& blocks = unit.blocks;
 
+  VLS env{unit, true};
+  auto vregConstraints = detail::computeDominatingConstraints(unit);
+
   // The vlower() implementations, and `lower', may allocate scratch blocks and
   // modify instruction streams, so we cannot use standard iterators here.
   PostorderWalker{unit}.dfs([&] (Vlabel b) {
     assertx(!blocks[b].code.empty());
 
+    env.allow_vreg = vregConstraints[b] != VregConstraint::Restrict;
```

Comment:
If we scope the restrictions, this can just be `env.allow_vreg = vregConstraint[b] == 0;`